`azurerm_container_app_environment`: Fix import diff on `workload_profile`

jiaweitao001 opened the pull request:
<!--  All Submissions -->


## Community Note
<!-- Please leave the community note as is. -->
* Please vote on this PR by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original PR to help the community and maintainers prioritize for review
* Please do not leave comments along the lines of "+1", "me too" or "any updates", they generate extra noise for PR followers and do not help prioritize for review


## Description

<!-- Please include a description below with the reason for the PR, what it is doing, what it is trying to accomplish, and anything relevant for a reviewer to know. 

If this is a breaking change for users please detail how it cannot be avoided and why it should be made in a minor version of the provider -->

Previously on 3.x, `consumption` workload profile was arbitrarily set. To avoid diff on the state file, if it's not explicitly configured we have a function to deliberately ignore it when reading.

In 4.x we do not have this scenario anymore. The `workload_profile` will always be the same as the user configured. The legacy logic will cause problem on importing existing resources because it ignores the `consumption` workload profile.


## PR Checklist

- [x] I have followed the guidelines in our [Contributing Documentation](../blob/main/contributing/README.md).
- [x] I have checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change.
- [x] I have checked if my changes close any open issues. If so please include appropriate [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) below.
- [x] I have updated/added Documentation as required written in a helpful and kind way to assist users that may be unfamiliar with the resource / data source.
- [x] I have used a meaningful PR title to help maintainers and other users understand this change and help prevent duplicate work. 
For example: “`resource_name_here` - description of change e.g. adding property `new_property_name_here`”


<!-- You can erase any parts of this template below this point that are not applicable to your Pull Request. -->


## Changes to existing Resource / Data Source

- [x] I have added an explanation of what my changes do and why I'd like you to include them (This may be covered by linking to an issue above, but may benefit from additional explanation).
- [x] I have written new tests for my resource or datasource changes & updated any relevant documentation.
- [x] I have successfully run tests with my changes locally. If not, please provide details on testing challenges that prevented you running the tests.
- [x] (For changes that include a **state migration only**). I have manually tested the migration path between relevant versions of the provider.


## Testing 

- [x] My submission includes Test coverage as described in the [Contribution Guide](../blob/main/contributing/topics/guide-new-resource.md) and the tests pass. (if this is not possible for any reason, please include details of why you did or could not add test coverage)

<!-- Please include testing logs or evidence here or an explanation on why no testing evidence can be provided. 

For state migrations please test the changes locally and provide details here, such as the versions involved in testing the migration path. For further details on testing state migration changes please see our guide on [state migrations](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/contributing/topics/guide-state-migrations.md#testing) in the contributor documentation. -->


## Change Log

Below please provide what should go into the changelog (if anything) conforming to the [Changelog Format documented here](../blob/main/contributing/topics/maintainer-changelog.md).

<!-- Replace the changelog example below with your entry. One resource per line. -->

* `azurerm_container_app_environment`: Fix import diff on `workload_profile`


<!-- What type of PR is this? -->
This is a (please select all that apply):

- [x] Bug Fix
- [ ] New Feature (ie adding a service, resource, or data source)
- [ ] Enhancement
- [ ] Breaking Change


## Related Issue(s)
Fixes #27057 #29383 


> [!NOTE] 
> If this PR changes meaningfully during the course of review please update the title and description as required.

miklosBartfai commented:
Please don't forget about this PR.
Now we can't do any modification on the CAE, because then it would get destroyed, with all the container apps running in it. It takes a lot of time (downtime) to redeploy everything, so we avoid it until we have no other choice.

mattaltberg commented:
bumping here, also a blocker in prod

catriona-m commented on internal/services/containerapps/container_app_environment_resource_test.go:
```diff
@@ -219,7 +219,7 @@ func TestAccContainerAppEnvironment_infraResourceGroup(t *testing.T) {
 			),
 		},
-		data.ImportStep(),
+		data.ImportStep("workload_profile"),
```
what's the reason for adding this as an importstep?

jiaweitao001 replied:
As described above.

catriona-m reviewed (changes requested):
Thanks for having another look at this @jiaweitao001. Could you explain your changes in the latest commit in more detail? In the original description we are removing the check for an explicitly defined consumption profile as it is no longer needed, but this change adds it back in? It would be good to get more clarity on what is happening here. Thanks!

jiaweitao001 commented:
> Thanks for having another look at this @jiaweitao001. Could you explain your changes in the latest commit in more detail? In the original description we are removing the check for an explicitly defined consumption profile as it is no longer needed, but this change adds it back in? It would be good to get more clarity on what is happening here. Thanks!

Sure. The origin of this confusion is from the behavior of the API. There are a total of 4 situations: 
1. `workload_profile` has both `consumption` and some other type set
2. `workload_profile` has only `consumtpion` set
3. `workload_profile` has only types other than `consumption` set
4. `workload_profile` is not configured

The API will automatically assign a `consumption` profile on situation 3. In my first draft of this PR, I did not realize it's still doing this, so I removed the check totally and break the corresponding acc tests. 

But if I simply put it back without change anything else, the import of existing resource with `consumption` set will break, just like the linked issue reported(during import the check will always have `false` value). That's why we need another check on non-consumption types.  If it is set and `consumption` profile is not, that means it fall on situation 3 I described above, then we ignore `consumption` profile during `read`, we can make sure there won't be any unexpected drift on state. For other situations as well as import, it will always read everything under `workload_profile`. 

As for the `ImportStep`, I added that because this change will surely break the import safety check. The imported values contains both `consumption` and non-consumption types.

catriona-m commented:
> > Thanks for having another look at this @jiaweitao001. Could you explain your changes in the latest commit in more detail? In the original description we are removing the check for an explicitly defined consumption profile as it is no longer needed, but this change adds it back in? It would be good to get more clarity on what is happening here. Thanks!
> 
> Sure. The origin of this confusion is from the behavior of the API. There are a total of 4 situations:
> 
> 1. `workload_profile` has both `consumption` and some other type set
> 2. `workload_profile` has only `consumtpion` set
> 3. `workload_profile` has only types other than `consumption` set
> 4. `workload_profile` is not configured
> 
> The API will automatically assign a `consumption` profile on situation 3. In my first draft of this PR, I did not realize it's still doing this, so I removed the check totally and break the corresponding acc tests.
> 
> But if I simply put it back without change anything else, the import of existing resource with `consumption` set will break, just like the linked issue reported(during import the check will always have `false` value). That's why we need another check on non-consumption types. If it is set and `consumption` profile is not, that means it fall on situation 3 I described above, then we ignore `consumption` profile during `read`, we can make sure there won't be any unexpected drift on state. For other situations as well as import, it will always read everything under `workload_profile`.
> 
> As for the `ImportStep`, I added that because this change will surely break the import safety check. The imported values contains both `consumption` and non-consumption types.

Thanks for explaining further @jiaweitao001 . Given that non-consumption profiles are not added by default, should we rely on what is returned from the API to determine the if there are non-consumption profiles rather than the state, which we won't have access to during import?

jiaweitao001 commented:
> > > Thanks for having another look at this @jiaweitao001. Could you explain your changes in the latest commit in more detail? In the original description we are removing the check for an explicitly defined consumption profile as it is no longer needed, but this change adds it back in? It would be good to get more clarity on what is happening here. Thanks!
> > 
> > 
> > Sure. The origin of this confusion is from the behavior of the API. There are a total of 4 situations:
> > 
> > 1. `workload_profile` has both `consumption` and some other type set
> > 2. `workload_profile` has only `consumtpion` set
> > 3. `workload_profile` has only types other than `consumption` set
> > 4. `workload_profile` is not configured
> > 
> > The API will automatically assign a `consumption` profile on situation 3. In my first draft of this PR, I did not realize it's still doing this, so I removed the check totally and break the corresponding acc tests.
> > But if I simply put it back without change anything else, the import of existing resource with `consumption` set will break, just like the linked issue reported(during import the check will always have `false` value). That's why we need another check on non-consumption types. If it is set and `consumption` profile is not, that means it fall on situation 3 I described above, then we ignore `consumption` profile during `read`, we can make sure there won't be any unexpected drift on state. For other situations as well as import, it will always read everything under `workload_profile`.
> > As for the `ImportStep`, I added that because this change will surely break the import safety check. The imported values contains both `consumption` and non-consumption types.
> 
> Thanks for explaining further @jiaweitao001 . Given that non-consumption profiles are not added by default, should we rely on what is returned from the API to determine the if there are non-consumption profiles rather than the state, which we won't have access to during import?

Hi @catriona-m , thank you for your suggestion. I think for import cases we cannot read from the API returned values. 

For import cases, we should import all `consumption` and non-consumption profiles, as you can see from the code change, in this case, both `consumptionDefined` and  `nonConsumptionDefined` will be `false`, no API returned values will be ignored. But if we read API returned value to determine if non-consumption profile is defined or not, we cannot tell the difference between this import case and a plain `read`, then we will fall in a dilemma of if we need to ignore the `consumption` profile or not.

miklosBartfai commented:
Just a reminder: this is still important, still a blocker, still waiting for the solution...

catriona-m commented:
Hi @jiaweitao001 - I looked into this a bit more to see if there was a way that we could suppress the diff in the case of having one default workload profile to work around this issue and came up with this [branch](https://github.com/hashicorp/terraform-provider-azurerm/compare/main...cm/workloadprofile). Is this something you could incorporate into this pr?  Or if you prefer, I could work on this and open a pr from the branch? WDYT?

jiaweitao001 commented:
> Hi @jiaweitao001 - I looked into this a bit more to see if there was a way that we could suppress the diff in the case of having one default workload profile to work around this issue and came up with this [branch](https://github.com/hashicorp/terraform-provider-azurerm/compare/main...cm/workloadprofile). Is this something you could incorporate into this pr? Or if you prefer, I could work on this and open a pr from the branch? WDYT?

Hi @catriona-m , thanks for sharing your thoughts here. I think your branch will also resolve this issue. Feel free to close mine and merge yours instead.
